timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/ArrayLockEventSlidingWindow.java

```diff
@@ -50,10 +57,29 @@ public OptionalLong getVersion() {
      */
     public synchronized void add(LockWatchEvent.Builder eventBuilder) {
         LockWatchEvent event = eventBuilder.build(nextSequence);
-        buffer[LongMath.mod(nextSequence, maxSize)] = event;
+        int headIndex = LongMath.mod(nextSequence, maxSize);
+        int tailIndex = LongMath.mod(lastVisibleSequence, maxSize);
+        if (headIndex == tailIndex) {
+            evictOldestEntryAtIndex(tailIndex);
+        }
+        buffer[headIndex] = event;
+        currentSize = currentSize + event.size();
+        if (currentSize > maxSize) {
+            int deletingIndex = LongMath.mod(lastVisibleSequence, maxSize);
+            while (currentSize > maxSize && lastVisibleSequence <= nextSequence - 1) {
+                evictOldestEntryAtIndex(deletingIndex);
+                deletingIndex = incrementAndMod(deletingIndex);
+            }
+        }
         nextSequence++;
     }
 
+    private void evictOldestEntryAtIndex(int index) {
+        currentSize = currentSize - buffer[index].size();
+        buffer[index] = PlaceholderLockWatchEvent.INSTANCE;
+        lastVisibleSequence++;
```

Comment:
Possibly better for offline discussion: Should this move up before we actually null out the entry?